File: backend/app/schemas/speaker_attribute_settings.py
"""Pydantic schemas for speaker attribute detection settings."""

from typing import Optional

from pydantic import BaseModel


class SpeakerAttributeSettings(BaseModel):
    """User-level speaker attribute settings."""

    detection_enabled: bool = True
    gender_detection_enabled: bool = True
    age_detection_enabled: bool = True
    show_attributes_on_cards: bool = True


class SpeakerAttributeSettingsUpdate(BaseModel):
    """Partial update for speaker attribute settings."""

    detection_enabled: Optional[bool] = None
    gender_detection_enabled: Optional[bool] = None
    age_detection_enabled: Optional[bool] = None
    show_attributes_on_cards: Optional[bool] = None


class SpeakerAttributeSystemDefaults(BaseModel):
    """System-level defaults for speaker attribute settings."""

    detection_enabled: bool
    gender_detection_enabled: bool
    age_detection_enabled: bool
    show_attributes_on_cards: bool

File: frontend/src/lib/api/speakerAttributeSettings.ts
/**
 * API service for speaker attribute detection settings
 *
 * Manages user preferences for speaker attribute detection including:
 * - Enable/disable attribute detection
 * - Gender prediction from voice
 * - Age range estimation from voice
 * - Display preferences for attribute badges on speaker cards
 */

import axiosInstance from '../axios';

/**
 * User speaker attribute detection settings
 */
export interface SpeakerAttributeSettings {
  detection_enabled: boolean;
  gender_detection_enabled: boolean;
  age_detection_enabled: boolean;
  show_attributes_on_cards: boolean;
}

/**
 * System-level defaults for speaker attribute detection
 */
export interface SpeakerAttributeSystemDefaults {
  detection_enabled: boolean;
  gender_detection_enabled: boolean;
  age_detection_enabled: boolean;
  show_attributes_on_cards: boolean;
}

/**
 * Get user's speaker attribute detection settings
 */
export async function getSpeakerAttributeSettings(): Promise<SpeakerAttributeSettings> {
  const response = await axiosInstance.get('/user-settings/speaker-attributes');
  return response.data;
}

/**
 * Update user's speaker attribute detection settings
 */
export async function updateSpeakerAttributeSettings(
  settings: Partial<SpeakerAttributeSettings>
): Promise<SpeakerAttributeSettings> {
  const response = await axiosInstance.put('/user-settings/speaker-attributes', settings);
  return response.data;
}

/**
 * Reset speaker attribute settings to system defaults
 */
export async function resetSpeakerAttributeSettings(): Promise<void> {
  await axiosInstance.delete('/user-settings/speaker-attributes');
}

/**
 * Get system default values for speaker attribute settings
 */
export async function getSpeakerAttributeSystemDefaults(): Promise<SpeakerAttributeSystemDefaults> {
  const response = await axiosInstance.get('/user-settings/speaker-attributes/system-defaults');
  return response.data;
}
